clients/orderbook/src/worker.rs

```diff
@@ -0,0 +1,1289 @@
+						// need to snapshot and drop everything else
+						self.snapshot(action.stid)?;
+						// We forget about everything else from cache.
+						self.known_messages.clear();
+						break
+					},
+				}
+			}
+			metric_set!(self, ob_state_id, last_snapshot);
+			last_snapshot = last_snapshot.saturating_add(1);
+		}
+		// We need to sub 1 since that last processed is one stid less than the not available
+		// when while loop is broken
+		self.last_snapshot.write().state_change_id = last_snapshot.saturating_sub(1);
+		Ok(())
+	}
+
+	pub fn want_stid(&mut self, from: &u64, to: &u64, peer: Option<PeerId>) {
+		if let Some(peer) = peer {
+			let mut messages = vec![];
+			for stid in *from..=*to {
+				// We dont allow gossip messsages to be greater than 10MB
+				if messages.encoded_size() >= 10 * 1024 * 1024 {
+					// If we reach size limit, we send data in chunks of 10MB.
+					let message = GossipMessage::Stid(messages);
+					self.gossip_engine.send_message(vec![peer], message.encode());
+					metric_inc!(self, ob_messages_sent);
+					metric_add!(self, ob_data_sent, message.encoded_size() as u64);
+					messages = vec![] // Reset the buffer
+				}
+				if let Some(msg) = self.known_messages.get(&stid) {
+					messages.push(msg.clone());
+				}
+			}
+			// Send the final chunk if any
+			if !messages.is_empty() {
+				let message = GossipMessage::Stid(messages);
+				self.gossip_engine.send_message(vec![peer], message.encode());
+				metric_inc!(self, ob_messages_sent);
+				metric_add!(self, ob_data_sent, message.encoded_size() as u64);
+			}
+		}
+	}
+
+	pub async fn got_stids_via_gossip(&mut self, messages: &Vec<ObMessage>) -> Result<(), Error> {
+		for message in messages {
+			// TODO: DO signature checks here and handle reputation
+			self.known_messages.entry(message.stid).or_insert(message.clone());
+		}
+		self.check_stid_gap_fill().await
+	}
+
+	// Expects the set bits in the bitmap to be missing chunks
+	pub async fn want(&mut self, snapshot_id: &u64, bitmap: &Vec<u128>, remote: Option<PeerId>) {
+		// Only respond if we are a fullnode
+		// TODO: Should we respond if we are also syncing???
+		if !self.is_validator {
+			if let Some(peer) = remote {
+				let mut chunks_we_have = vec![];
+				let at = BlockId::Number(self.last_finalized_block.saturated_into());
+				if let Ok(Some(summary)) =
+					self.runtime.runtime_api().get_snapshot_by_id(&at, *snapshot_id)
+				{
+					if let Some(offchain_storage) = self.backend.offchain_storage() {
+						let required_indexes: Vec<u16> = return_set_bits(bitmap);
+						for chunk_index in required_indexes {
+							if offchain_storage
+								.get(
+									ORDERBOOK_STATE_CHUNK_PREFIX,
+									summary.state_chunk_hashes[chunk_index as usize]
+										.encode()
+										.as_ref(),
+								)
+								.is_some()
+							{
+								chunks_we_have.push(chunk_index);
+							}
+						}
+					}
+				} else {
+					// TODO: Reduce reputation if else block is happens
```

Comment:
Still pending

Reply:
Reputation is not part of this version but these are places we need to handle it in next version.